govcd/nsxt_distributed_firewall.go

```diff
@@ -99,3 +109,329 @@ func (firewall *DistributedFirewall) DeleteAllRules() error {
+	}
+
+	client := vdcGroup.client
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointVdcGroupsDfwRules
+	apiVersion, err := client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// "default" policy is hardcoded because there is no other policy supported
+	urlRef, err := client.OpenApiBuildEndpoint(fmt.Sprintf(endpoint, vdcGroup.VdcGroup.Id, types.DistributedFirewallPolicyDefault), "/", id)
+	if err != nil {
+		return nil, err
+	}
+
+	returnObject := &DistributedFirewallRule{
+		Rule:     &types.DistributedFirewallRule{},
+		client:   client,
+		VdcGroup: vdcGroup,
+	}
+
+	err = client.OpenApiGetItem(apiVersion, urlRef, nil, returnObject.Rule, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving Distributed Firewall rule: %s", err)
+	}
+
+	return returnObject, nil
+}
+
+// GetDistributedFirewallRuleByName retrieves single firewall rule by name
+func (vdcGroup *VdcGroup) GetDistributedFirewallRuleByName(name string) (*DistributedFirewallRule, error) {
+	if name == "" {
+		return nil, fmt.Errorf("name must be specified")
+	}
+
+	dfw, err := vdcGroup.GetDistributedFirewall()
+	if err != nil {
+		return nil, fmt.Errorf("error returning distributed firewall rules: %s", err)
+	}
+
+	var filteredByName []*types.DistributedFirewallRule
+	for _, rule := range dfw.DistributedFirewallRuleContainer.Values {
+		if rule.Name == name {
+			filteredByName = append(filteredByName, rule)
+		}
+	}
+
+	oneByName, err := oneOrError("name", name, filteredByName)
+	if err != nil {
+		return nil, err
+	}
+
+	return vdcGroup.GetDistributedFirewallRuleById(oneByName.ID)
+}
+
+// CreateDistributedFirewallRule is a wrapper around "vdcGroups/%s/dfwPolicies/%s/rules" endpoint
+// which handles all distributed firewall (dfw) rules at once. While there is no real endpoint to
+// create single firewall rule, it is a requirements for some cases (e.g. using in Terraform)
+// The code works by doing the following steps:
+//
+// 1. Getting all Distributed Firewall Rules and storing them in private intermediate
+// type`distributedFirewallRulesRaw` which holds a []json.RawMessage (text) instead of exact types.
+// This will prevent altering existing rules in any way (for example if a new field appears in
+// schema in future VCD versions)
+//
+// 2. Converting the give `rule` into json.RawMessage so that it is provided in the same format as
+// other already retrieved rules
+//
+// 3. Creating a new structure of []json.RawMessage which puts the new rule into one of places:
+// 3.1. to the end of []json.RawMessage - bottom of the list
+// 3.2. if `optionalAboveRuleId` argument is specified - identifying the position and placing new
+// rule above it
+// 4. Perform a PUT (update) call to the "vdcGroups/%s/dfwPolicies/%s/rules" endpoint using the
+// newly constructed payload
+//
+// Note. Running this function concurrently will corrupt firewall rules as it uses an endpoint that
```

Comment:
I was waiting to read something like this :) I believe it is very risky to leave concurrency handling to the caller of this function. The cost is too high in such case, as it would lead to data corruption, which we cannot afford. What about synchronizing this function right here, so the callers don't have to?

FYI: @dataclouder @adambarreiro

Reply:
I am not sure we want to do an exception here. There are plenty of situations where one can hit such problems (https://github.com/vmware/go-vcloud-director/pull/181)
But back then we agreed to offload it to consumer (Terraform provider in our case)
https://github.com/vmware/go-vcloud-director/issues/180#issuecomment-547390216


OTOH, from the technical side I don't see a big problem making a locking mechanism, but it is just not something that we have done in the SDK (yet)

Reply:
After leaving my comment, I started thinking about that as well...

Would you share a link where is the locking for this rule management implemented currently? For some reason, I'm having a hard time quickly locating it. Thanks!

Reply:
Both `vcd_distributed_firewall` and `vcd_distributed_firewall_rule` are locking on parent VDC Group. Only one operation can happen at a time at that particular VDC Group. (There can be only one Distributed Firewall per VDC Group {at least for now} )

* https://github.com/vmware/terraform-provider-vcd/pull/1076/files#diff-fb43a12d1b2bebd9d74f4c764d71fa5d5886715209c589c5e385d1ce12c8ac07R143-R144
* https://github.com/vmware/terraform-provider-vcd/blob/main/vcd/resource_vcd_nsxt_distributed_firewall.go#L154-L156

Reply:
I think the note is good enough. Possible suggestions could be

```
// CreateDistributedFirewallRule is a non-thread safe wrapper ...
```

or

```
// WARNING: Running this function ...
```

Just to make the "heads up" more visible.

Reply:
@adambarreiro once you're happy with the note above, please feel free to resolve this conversation, as my questions have been answered here :)

Reply:
> PS: I'd choose the first one, you can get the "warning" that is not thread-safe in a glimpse.

Is it the first one, or second with WARNING?

Reply:
I'd go with `// CreateDistributedFirewallRule is a non-thread safe wrapper `

Reply:
> I'd go with // CreateDistributedFirewallRule is a non-thread safe wrapper

Made it so